### PR TITLE
preserve multi-select values for dynamic questions when the list of choices change

### PR DESCRIPTION
**Pull request by snopoke**
## Product Description
This fixes a bug in Web Apps where changes in the choices for lookup table questions cause the selections to get cleared.

With this change the selections are preserved (where possible) when the options change.

Peek 2021-12-09 16-35

## Technical Summary
Prior to this change the UI code uses array indexes as the selected value but using indexes is problematic because the meaning of an index can change if the list of choices change which makes it impossible to safely preserve selections if options change.

Formplayer does already update the answer index when the list of choices change so the only issue is on the Web Apps UI side. Changing to using the actual option text makes if possible to preserve the selection even when the options change. 

This PR also changes the widget template code to use standard Knockout bindings for the options instead of creating them manually (and therefore also changes the `SingleSelectEntry` which uses the same template.

Jira: https://dimagi-dev.atlassian.net/browse/USH-1276

### Safety story
This could adversely affect how the checkbox and multi-select questions work. QA is going to be done on staging. There are also some test cases that cover the changes.

### Automated test coverage
There are some tests for these classes. I've updated them and added some more.

### QA Plan
QA ticket to follow

### Migrations
NA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by orangejenny** [comment]
Didn't get to this today, should be able to review tomorrow.

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js [hunk 1]
```diff
@@ -472,27 +472,19 @@ hqDefine("cloudcare/js/form_entry/entries", function () {
...
-                self.onPreProcess.call(this, answer === 0 ? Const.NO_ANSWER : answer);
+                let answer = _.indexOf(choices, value);
+                this.answer(answer === -1 ? Const.NO_ANSWER : answer + 1);
             }
         });
 
         self.onClear = function () {
             self.rawAnswer(Const.NO_ANSWER);
         };
-        self.isValid = function () {
```
Comment (orangejenny) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js:457]: nice simplification

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/entries_spec.js [hunk 2]
```diff
@@ -228,6 +228,24 @@ describe('Entries', function () {
         assert.sameMembers(entry.answer(), [2]);
     });
 
+    it('Should retain MultiSelectEntry when choices change', function () {
+        questionJSON.datatype = Const.MULTI_SELECT;
+        questionJSON.choices = ['a', 'b'];
+        questionJSON.answer = [1, 2]; // answer is based on a 1 indexed index of the choices
```
Comment (orangejenny) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/entries_spec.js:234]: A comment like this with where the new `rawAnswer` computeds are defined would be helpful.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js [hunk 3]
```diff
@@ -451,26 +459,32 @@ hqDefine("cloudcare/js/form_entry/entries", function () {
...
+                    return Const.NO_ANSWER;
+                }
+
+                let choices = this.choices();
+                return choices[answer - 1];
+            },
+            write: (value) => {
+                let choices = this.choices.peek();
+                let answer = _.indexOf(choices, value);
+                this.answer(answer === -1 ? Const.NO_ANSWER : answer + 1);
```
Comment (snopoke) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js:476]:
```suggestion
                // answer is based on a 1 indexed index of the choices
                this.answer(answer === -1 ? Const.NO_ANSWER : answer + 1);
```

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js [hunk 4]
```diff
@@ -397,6 +397,23 @@ hqDefine("cloudcare/js/form_entry/entries", function () {
...
+                if (answer === Const.NO_ANSWER) {
+                    return [];
+                }
+
+                let choices = this.choices();
+                return answer.map(index => choices[index - 1]);
+            },
+            write: (value) => {
+                let choices = this.choices.peek();
+                let answer = _.filter(value.map((val) => _.indexOf(choices, val) + 1), (v) => v > 0);
```
Comment (snopoke) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js:412]:
```suggestion
                // answer is based on a 1 indexed index of the choices
                let answer = _.filter(value.map((val) => _.indexOf(choices, val) + 1), (v) => v > 0);
```

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js [hunk 5]
```diff
@@ -451,26 +460,33 @@ hqDefine("cloudcare/js/form_entry/entries", function () {
...
+
+                let choices = this.choices();
+                return choices[answer - 1];
+            },
+            write: (value) => {
+                let choices = this.choices.peek();
+                let answer = _.indexOf(choices, value);
+                // answer is based on a 1 indexed index of the choices
+                this.answer(answer === -1 ? Const.NO_ANSWER : answer + 1);
+            }
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/cloudcare/static/cloudcare/js/form_entry/entries.js:479]:
```suggestion
            },
```

**Pull request merged**